cmd/controller/main.go

```diff
@@ -53,5 +71,55 @@ func main() {
 		"reconciliation-timeout", reconciler.DefaultTimeout,
 		"The amount of time to give each reconciliation of a resource to complete before its context is canceled.")
 
-	sharedmain.MainWithContext(signals.NewContext(), "controller", ctors...)
+	ctx := signals.NewContext()
+
+	// Allow configuration of threads per controller
+	if val, ok := os.LookupEnv("K_THREADS_PER_CONTROLLER"); ok {
+		threadsPerController, err := strconv.Atoi(val)
+		if err != nil {
+			log.Fatalf("failed to parse value %q of K_THREADS_PER_CONTROLLER: %v\n", val, err)
+		}
+		controller.DefaultThreadsPerController = threadsPerController
+	}
+
+	// TODO(mattmoor): Remove this once HA is stable.
+	disableHighAvailability := flag.Bool("disable-ha", false,
+		"Whether to disable high-availability functionality for this component.  This flag will be deprecated "+
+			"and removed when we have promoted this feature to stable, so do not pass it without filing an "+
+			"issue upstream!")
+
+	// HACK: This parses flags, so the above should be set once this runs.
+	cfg := injection.ParseAndGetRESTConfigOrDie()
+
+	if *disableHighAvailability {
+		ctx = sharedmain.WithHADisabled(ctx)
+	}
+
+	// If nil it panics
+	client := kubernetes.NewForConfigOrDie(cfg)
+
+	if shouldEnableNetCertManagerController(ctx, client) {
+		injection.Default.RegisterInformer(challenge.WithInformer)
+		injection.Default.RegisterInformer(v1certificate.WithInformer)
+		injection.Default.RegisterInformer(certificaterequest.WithInformer)
+		injection.Default.RegisterInformer(clusterissuer.WithInformer)
+		injection.Default.RegisterInformer(issuer.WithInformer)
+		ctors = append(ctors, netcertmanager.NewController)
+	}
+
+	sharedmain.MainWithConfig(ctx, "controller", cfg, ctors...)
+}
+
+func shouldEnableNetCertManagerController(ctx context.Context, client *kubernetes.Clientset) bool {
+	var cm *v1.ConfigMap
+	var err error
+	if cm, err = client.CoreV1().ConfigMaps(system.Namespace()).Get(ctx, "config-network", metav1.GetOptions{}); err != nil {
+		log.Fatalf("Failed to get cm config-network: %v", err)
+	}
+	netCfg, err := netcfg.NewConfigFromMap(cm.Data)
+	if err != nil {
+		log.Fatalf("Failed to construct network config: %v", err)
+	}
+	return netCfg.ExternalDomainTLS || netCfg.SystemInternalTLSEnabled() || (netCfg.ClusterLocalDomainTLS == netcfg.EncryptionEnabled) ||
```

Comment:
We should validate cert manager CRDs as installed

Reply:
I can do that but also the user could know via docs that internal encryption requires cert manager. Personally I think it is a bit too much but I can add it.

Reply:
done.